pkg/controller/runtime/runtime.go

```diff
@@ -240,32 +246,133 @@ func (runtime *Runtime) watch(resourceNamespace resource.Namespace, resourceType
+	// Perform deduplication of events based on the reduction of the event value to the reducedMetadata.
+	//
+	// deduplication process consists of two goroutines:
+	// 1. the first goroutine reads events from the watch channel as fast as possible,
+	//    reduces the event value to 'reducedMetadata' and pushes them to the map
+	// 2. the second goroutine consumes a map from the first goroutine, performs the work required
+	//    to trigger updates in the dependent controller
+	//
+	// The design idea is to consume watch events as fast as possible, while delaying "heavy" work
+	// to the second goroutine.
+	//
+	// There is a trick being used which sends a single map back and forth between the two goroutines.
+	// There is no locking required, as the map is owned by a single goroutine at a single moment of time.
+	// Additional channel 'empty' is used to block the second goroutine when there are no events to process.
+	ch := make(chan dedup, 1)
+	empty := make(chan dedup, 1)
+	empty <- dedup{}
+
+	go runtime.deduplicateWatchEvents(ch, empty)
+	go runtime.deliverDeduplicatedEvents(ch, empty)
+}
+
+// deduplicateWatchEvents deduplicates events from the watch channel into the map sent to the channel ch.
+func (runtime *Runtime) deduplicateWatchEvents(ch chan dedup, empty <-chan dedup) {
 	for {
-		var e state.Event
+		var events []state.Event
+
+		// wait for an event
+		select {
+		case <-runtime.runCtx.Done():
+			return
+		case events = <-runtime.watchCh:
+		}
+
+		// acquire a map
+		var m dedup
 
 		select {
+		case m = <-empty:
+		case m = <-ch:
 		case <-runtime.runCtx.Done():
 			return
-		case e = <-runtime.watchCh:
 		}
 
-		if e.Type == state.Errored {
-			// watch failed, we need to abort
-			runtime.watchErrors <- e.Error
+		processEvents := func() bool {
+			for _, e := range events {
+				if e.Type == state.Errored {
+					// watch failed, we need to abort
+					runtime.watchErrors <- e.Error
+
+					return false
+				}
+
+				m[reduceMetadata(e.Resource.Metadata())] = struct{}{}
+			}
 
+			return true
+		}
+
+		if !processEvents() {
 			return
 		}
 
-		md := e.Resource.Metadata()
+		// drain the watchCh by consuming all immediately available events
+	drainer:
+		for {
+			select {
+			case events = <-runtime.watchCh:
+				if !processEvents() {
+					return
+				}
+			case <-runtime.runCtx.Done():
+				return
+			default:
+				break drainer
+			}
+		}
+
+		// send the map to the second goroutine for processing
+		if !channel.SendWithContext(runtime.runCtx, ch, m) {
+			return
+		}
+	}
+}
+
+// deliverDeduplicatedEvents delivers events from the deduplicated channel to the controllers.
+func (runtime *Runtime) deliverDeduplicatedEvents(ch chan dedup, empty chan<- dedup) {
+	for {
+		// wait for a map
+		var m dedup
+		select {
+		case m = <-ch:
+		case <-runtime.runCtx.Done():
+			return
+		}
+
+		// consume any first key of the map
+		var k reducedMetadata
+
+		for k = range m {
+			break
+		}
+
+		delete(m, k)
```

Comment:
I think we can just add `takeOne` method to `dedup` type, since it is custom type anyway.

Something like:
```go
func (d dedup) takeOne() reducedMetadata {
	for k := range d {
		delete(d, k)
		return k
	}
}
```

```

Reply:
fixed